src/main/java/org/myrobotlab/framework/Service.java

```diff
@@ -805,13 +803,58 @@ public String getDataInstanceDir() {
     return Runtime.getOptions().dataDir + fs + getClass().getSimpleName() + fs + getName();
   }
 
+  public String getResourceRoot() {
+    // FIXME - should "this" be the test ?
+    // If so it should be its own static function...
+
+    // order of precedence
+    // 1. if there is a src directory use it unless
+    // 2. options say to override it
+
+    String resourceRoot = Runtime.getOptions().resourceDir;
+    if ("resource".equals(resourceRoot)) {
+      // allow default to be overriden by src if it exists
+      File src = new File("src");
+      if (src.exists()) {
+        resourceRoot = "src" + fs + "main" + fs + "resources" + fs + "resource";
+      }
+    }
+
+    return resourceRoot;
+  }
+
   public String getResourceDir() {
-    String dataDir = Runtime.getOptions().resourceDir + fs + getClass().getSimpleName();
-    File f = new File(dataDir);
+
+    String resourceDir = getResourceRoot() + fs + getClass().getSimpleName();
+    File f = new File(resourceDir);
     if (!f.exists()) {
       f.mkdirs();
     }
-    return Runtime.getOptions().resourceDir + fs + getClass().getSimpleName();
+    return resourceDir;
+  }
+
+  public byte[] getResource(String resourceName) {
+    String filename = getResourceDir() + fs + resourceName;
+    File f = new File(filename);
+    if (!f.exists()) {
+      error("resource %s does not exist", f);
+      return null;
+    }
+    byte[] content = null;
+    try {
+      content = Files.readAllBytes(Paths.get(filename));
+    } catch (IOException e) {
+      error(e);
+    }
+    return content;
+  }
+
+  public String getResourceAsString(String resourceName) {

```

Comment:
this method should be explicit about the encoding that the resource will be decoded as.  ( utf-8? )  or the system default?  I worry about internationalization issues with this method.  we should force utf-8  (unless for some reason we don't want to do that?)

Reply:
sure .. fixed